arch/network_tcp.c

```diff
@@ -25,6 +25,142 @@
+    UA_initialize_architecture_network();
+
+    ServerNetworkLayerTCP *layer = (ServerNetworkLayerTCP *)nl->handle;
+    layer->logger = logger;
+
+    /* Get addrinfo of the server and create server sockets */
+    char hostname[512];
+    if(customHostname->length) {
+        if(customHostname->length >= sizeof(hostname))
+            return UA_STATUSCODE_BADOUTOFMEMORY;
+        memcpy(hostname, customHostname->data, customHostname->length);
+        hostname[customHostname->length] = '\0';
+    }
+    char portno[6];
+    UA_snprintf(portno, 6, "%d", layer->port);
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof hints);
+
+    /* Get the discovery url from the hostname */
+    UA_String du = UA_STRING_NULL;
+    char discoveryUrlBuffer[256];
+    if(customHostname->length) {
+        du.length = (size_t)UA_snprintf(discoveryUrlBuffer, 255, "opc.tcp://%.*s:%d/",
+                                        (int)customHostname->length, customHostname->data,
+                                        layer->port);
+        du.data = (UA_Byte*)discoveryUrlBuffer;
+    } else {
+        char hostnameBuffer[256];
+        if(UA_gethostname(hostnameBuffer, 255) == 0) {
+            du.length = (size_t)UA_snprintf(discoveryUrlBuffer, 255, "opc.tcp://%s:%d/",
+                                            hostnameBuffer, layer->port);
+            du.data = (UA_Byte*)discoveryUrlBuffer;
+        } else {
+            UA_LOG_ERROR(layer->logger, UA_LOGCATEGORY_NETWORK, "Could not get the hostname");
+            return UA_STATUSCODE_BADINTERNALERROR;
+        }
+    }
+    UA_String_copy(&du, &nl->discoveryUrl);
+
+    UA_LOG_INFO(layer->logger, UA_LOGCATEGORY_NETWORK,
+                "TCP network layer listening on %.*s",
+                (int)nl->discoveryUrl.length, nl->discoveryUrl.data);
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+ServerNetworkLayerTCP_listen(UA_ServerNetworkLayer *nl, UA_Server *server,
+                             UA_UInt16 timeout) {
+    return UA_STATUSCODE_GOOD;
+}
+
+static void
+ServerNetworkLayerTCP_stop(UA_ServerNetworkLayer *nl, UA_Server *server) {
+    UA_deinitialize_architecture_network();
+}
+
+/* run only when the server is stopped */
+static void
+ServerNetworkLayerTCP_clear(UA_ServerNetworkLayer *nl) {
+    ServerNetworkLayerTCP *layer = (ServerNetworkLayerTCP *)nl->handle;
+    UA_String_clear(&nl->discoveryUrl);
+
+    /* Hard-close and remove remaining connections. The server is no longer
+     * running. So this is safe. */
+    ConnectionEntry *e, *e_tmp;
+    LIST_FOREACH_SAFE(e, &layer->connections, pointers, e_tmp) {
+        LIST_REMOVE(e, pointers);
+        layer->connectionsSize--;
+        UA_free(e);
+        if(nl->statistics) {
+            nl->statistics->currentConnectionCount--;
+        }
+    }
+
+    /* Free the layer */
+    UA_free(layer);
+}
+
+UA_ServerNetworkLayer
+UA_ServerNetworkLayerTCP(UA_ConnectionConfig config, UA_UInt16 port,
```

Comment:
This is no longer used, right?
So it can be deleted.

Reply:
It is still used, unfortunately the layers and their internal structures that is filled with information are not that loosely coupled to some of the internal functionalities, so for breaking as little as possible of some layer related code (and for changing as little as possible), they will still be created and removed serving as pure storage of information.

Reply:
It is used in the function `UA_ServerConfig_addNetworkLayerTCP` in `plugins/ua_config_default.c` ll. 331 ff when the networklayers get set up:

```c
    conf->networkLayers[conf->networkLayersSize] =
        UA_ServerNetworkLayerTCP(config, portNumber, 0);
    if (!conf->networkLayers[conf->networkLayersSize].handle)
        return UA_STATUSCODE_BADOUTOFMEMORY;
    conf->networkLayersSize++;

```

Reply:
Yes. But you never use it.
It just sits there doing nothing.